### PR TITLE
Update wb_presets.c (Nikon Coolpix P1000)

### PR DESCRIPTION
**Pull request by Macchiato17**
Adding the whitebalance values for Nikon Coolpix P1000, firmware 1.2.

I extracted the values with the help of your Ruby script extract_wb on Windows 10 professional x64 as described on https://www.bilddateien.de/blog/2019-01-20-kamera-weissabgleichseinstellungen-fuer-darktable-ermitteln.html. 

When running the script, there was an error about something like "array not being an array" in line 164 (found_presets = Hash[found_presets.sort.zip].keys), so I commented this line and manually did the sorting and removal of duplicates in Notepad++. After that, there were some occurrences of value-duplicates (in the curly braces) with different WB-tags assigned, which I also removed from the list. If you need the original (unsorted) list, please feel free to send me an e-mail :-)

I'm wondering, why the whitebalancing shifting-values (e.g. -3...0...+3) which I applied where possible, is not found in the resulting list anymore.

### CODE REVIEW

File: src/external/wb_presets.c [hunk 1]
```diff
@@ -2350,6 +2350,74 @@ const wb_data wb_preset[] =
   { "Nikon", "Coolpix P340", Cloudy, 0,			{ 1.976563, 1, 1.457031, 0 } },
   { "Nikon", "Coolpix P340", Flash, 0,			{ 2.085938, 1, 1.554688, 0 } },
 
+  // Nikon CoolPix P1000 firmware 1.2
+  { "Nikon", "COOLPIX P1000", HTMercury           , 0, { 1.76953125, 1, 1.61328125, 0 } },
```
Comment (TurboGit) [inline src/external/wb_presets.c]: HTMercury is unknown. You may want to add it on top of the file in the Nikon section to allow translation.

**Comment by Macchiato17** [comment]
Thanks for the quick replay, I hope I added the definition the right way.

**Comment by TurboGit** [comment]
Looks ok, merged now. Thanks.

File: src/external/wb_presets.c [hunk 2]
```diff
@@ -2350,6 +2352,74 @@ const wb_data wb_preset[] =
...
+  { "Nikon", "COOLPIX P1000", DirectSunlight      , 0, { 1.4375, 1, 1.99609375, 0 } },
+  { "Nikon", "COOLPIX P1000", DirectSunlight      , 0, { 1.46875, 1, 1.94140625, 0 } },
+  { "Nikon", "COOLPIX P1000", DirectSunlight      , 0, { 1.50390625, 1, 1.88671875, 0 } },
+  { "Nikon", "COOLPIX P1000", DirectSunlight      , 0, { 1.5390625, 1, 1.8203125, 0 } },
+  { "Nikon", "COOLPIX P1000", DirectSunlight      , 0, { 1.578125, 1, 1.7578125, 0 } },
+  { "Nikon", "COOLPIX P1000", DirectSunlight      , 0, { 1.61328125, 1, 1.6953125, 0 } },
+  { "Nikon", "COOLPIX P1000", DirectSunlight      , 0, { 1.40625, 1, 2.05078125, 0 } },
```
Comment (johnny-bit) [inline src/external/wb_presets.c]: This causes finetuning for `DirectSunlight` to be not taken into consideration!
Also this is weird - it looks like you had -3..-2..-1..0..1..2..-4 based off of R&B coefficients.

Can you regenerate those?

File: src/external/wb_presets.c [hunk 3]
```diff
@@ -2350,6 +2352,74 @@ const wb_data wb_preset[] =
...
+  { "Nikon", "COOLPIX P1000", Daylight            , 0, { 1.50390625, 1, 1.88671875, 0 } },
+  { "Nikon", "COOLPIX P1000", Shade               , 0, { 1.87890625, 1, 1.4140625, 0 } },
+  { "Nikon", "COOLPIX P1000", Cloudy              , 0, { 1.61328125, 1, 1.765625, 0 } },
+  { "Nikon", "COOLPIX P1000", Cloudy              , 0, { 1.65234375, 1, 1.71875, 0 } },
+  { "Nikon", "COOLPIX P1000", Cloudy              , 0, { 1.734375, 1, 1.6015625, 0 } },
+  { "Nikon", "COOLPIX P1000", Cloudy              , 0, { 1.7734375, 1, 1.546875, 0 } },
+  { "Nikon", "COOLPIX P1000", Cloudy              , 0, { 1.578125, 1, 1.8203125, 0 } },
+  { "Nikon", "COOLPIX P1000", Cloudy              , 0, { 1.54296875, 1, 1.87109375, 0 } },
+  { "Nikon", "COOLPIX P1000", Cloudy              , 0, { 1.69140625, 1, 1.66015625, 0 } },
```
Comment (johnny-bit) [inline src/external/wb_presets.c]: This causes finetuning for `Cloudy` to be not taken into consideration! ALL finetuning values are `0`!

Also there's something wrong with them and seem to be processed out of order. Just take a look at R coefficient it goes from 1.61 to 1.77 then jumps to 1.57 and down to 1.54 and suddenly back to 1.69.

File: src/external/wb_presets.c [hunk 4]
```diff
@@ -2350,6 +2352,74 @@ const wb_data wb_preset[] =
...
+  { "Nikon", "COOLPIX P1000", Tungsten            , 0, { 1.015625, 1, 2.8046875, 0 } },
+  { "Nikon", "COOLPIX P1000", Incandescent        , 0, { 0.9921875, 1, 2.88671875, 0 } },
+  { "Nikon", "COOLPIX P1000", Incandescent        , 0, { 0.96875, 1, 2.96875, 0 } },
+  { "Nikon", "COOLPIX P1000", Incandescent        , 0, { 1.015625, 1, 2.8046875, 0 } },
+  { "Nikon", "COOLPIX P1000", Incandescent        , 0, { 1.0625, 1, 2.6171875, 0 } },
+  { "Nikon", "COOLPIX P1000", Incandescent        , 0, { 1.0390625, 1, 2.7109375, 0 } },
+  { "Nikon", "COOLPIX P1000", Incandescent        , 0, { 0.94921875, 1, 3.05078125, 0 } },
+  { "Nikon", "COOLPIX P1000", Incandescent        , 0, { 1.08984375, 1, 2.5234375, 0 } },
```
Comment (johnny-bit) [inline src/external/wb_presets.c]: This causes finetuning for `Incadescent` to be not taken into consideration. Again, all values for finetuning are `0`.

Also there are another weird coefficeint jums so we can't be sure in what order are those.

File: src/external/wb_presets.c [hunk 5]
```diff
@@ -2350,6 +2352,74 @@ const wb_data wb_preset[] =
...
+  { "Nikon", "COOLPIX P1000", CoolWhiteFluorescent, 0, { 1.5859375, 1, 2.5, 0 } },
+  { "Nikon", "COOLPIX P1000", NeutralFluorescent  , 0, { 1.59765625, 1, 1.83203125, 0 } },
+  { "Nikon", "COOLPIX P1000", DaylightFluorescent , 0, { 1.796875, 1, 1.47265625, 0 } },
+  { "Nikon", "COOLPIX P1000", Flash               , 0, { 1.8984375, 1, 1.44921875, 0 } },
+  { "Nikon", "COOLPIX P1000", Flash               , 0, { 1.8125, 1, 1.55859375, 0 } },
+  { "Nikon", "COOLPIX P1000", Flash               , 0, { 1.65234375, 1, 1.75390625, 0 } },
+  { "Nikon", "COOLPIX P1000", Flash               , 0, { 1.69140625, 1, 1.70703125, 0 } },
+  { "Nikon", "COOLPIX P1000", Flash               , 0, { 1.73046875, 1, 1.66015625, 0 } },
+  { "Nikon", "COOLPIX P1000", Flash               , 0, { 1.76953125, 1, 1.61328125, 0 } },
+  { "Nikon", "COOLPIX P1000", Flash               , 0, { 1.85546875, 1, 1.50390625, 0 } },
```
Comment (johnny-bit) [inline src/external/wb_presets.c]: This again has no finetuning applicable to it and has weird coefficient jumps so it's hard to know which one is actually 0 and if they are in -3..3 or 3..-3 order :/

**Comment by johnny-bit** [comment]
I resorted those and fixed in #4548. Do note though that I don't have Nikon P1000 to test it, but by judging the coefficients it should be ok.

**Comment by Macchiato17** [comment]
Hi Hubert, thanks a lot for all your information and effort. I will also check on my side hopefully this week. I still have the pictures and mapping to see which setting applies to which photo. Bye, Lars.
________________________________

**Pull request merged**